addons/hr_holidays/models/hr_holidays.py

```diff
@@ -0,0 +1,420 @@
+                    leaves.append(self.with_context(mail_notify_force_send=False).create(vals))
+                for leave in leaves:
+                    # TODO is it necessary to interleave the calls?
+                    for sig in ('confirm', 'validate', 'second_validate'):
+                        leave.signal_workflow(sig)
+        return True
+
+    @api.multi
+    def holidays_confirm(self):
+        return self.write({'state': 'confirm'})
+
+    @api.multi
+    def holidays_refuse(self):
+        manager = self.env['hr.employee'].search([('user_id', '=', self.env.uid)], limit=1)
+        for holiday in self:
+            if holiday.state == 'validate1':
+                holiday.write({'state': 'refuse', 'manager_id': manager.id})
+            else:
+                holiday.write({'state': 'refuse', 'manager_id2': manager.id})
+        self.holidays_cancel()
+        return True
+
+    @api.multi
+    def holidays_cancel(self):
+        for holiday in self:
+            # Delete the meeting
+            if holiday.meeting_id:
+                holiday.meeting_id.unlink()
+
+            # If a category that created several holidays, cancel all related
+            holiday.linked_request_ids.signal_workflow('refuse')
+
+        self._remove_resource_leave()
+        return True
+
+    @api.multi
+    def toggle_payslip_status(self):
+        for holiday in self:
+            holiday.payslip_status = not holiday.payslip_status
+        return True
+
+    @api.multi
+    def _track_subtype(self, init_values):
+        self.ensure_one()
+        if 'state' in init_values and self.state == 'validate':
+            return 'hr_holidays.mt_holidays_approved'
+        elif 'state' in init_values and self.state == 'validate1':
+            return 'hr_holidays.mt_holidays_first_validated'
+        elif 'state' in init_values and self.state == 'confirm':
+            return 'hr_holidays.mt_holidays_confirmed'
+        elif 'state' in init_values and self.state == 'refuse':
+            return 'hr_holidays.mt_holidays_refused'
+        return super(HrHolidays, self)._track_subtype(init_values)
+
+    @api.multi
+    def _notification_group_recipients(self, message, recipients, done_ids, group_data):
+        """ Override the mail.thread method to handle HR users and officers
+        recipients. Indeed those will have specific action in their notification
+        emails. """
+        recipients_to_add = recipients.filtered(lambda r: r.id not in done_ids and r.user_ids and r.user_ids[0].has_group('base.group_hr_user'))
+        group_data['group_hr_user'] |= recipients_to_add
+        done_ids.update(set(recipients_to_add.ids))
+        return super(HrHolidays, self)._notification_group_recipients(message, recipients, done_ids, group_data)
+
+    @api.multi
+    def _notification_get_recipient_groups(self, message, recipients):
+        res = super(HrHolidays, self)._notification_get_recipient_groups(message, recipients)
+
+        holiday = self[0]
```

Comment:
Why self[0] ?. Use self.ensure_one() to verify that the self contains only one record.